pkg/lightning/backend/external/util.go

```diff
@@ -128,56 +128,74 @@ func seekPropsOffsets(
 	return offsetsPerKey, nil
 }
 
-// GetAllFileNames returns data file paths and stat file paths. Both paths are
-// sorted.
+// GetAllFileNames returns files with the same non-partitioned dir.
+//   - for intermediate KV/stat files we store them with a partitioned way to mitigate
+//     limitation on Cloud, see randPartitionedPrefix for how we partition the files.
+//   - for meta files, we store them directly under the non-partitioned dir.
+//
+// for example, if nonPartitionedDir is '30001', the files returned might be
+//   - 30001/6/meta.json
+//   - 30001/7/meta.json
+//   - 30001/plan/ingest/1/meta.json
+//   - 30001/plan/merge-sort/1/meta.json
+//   - e6/30001/7/617527bf-e25d-4312-8784-4a4576eb0195_stat/one-file
+//   - fa/30001/7/617527bf-e25d-4312-8784-4a4576eb0195/one-file
 func GetAllFileNames(
 	ctx context.Context,
 	store storage.ExternalStorage,
-	subDir string,
-) ([]string, []string, error) {
+	nonPartitionedDir string,
+) ([]string, error) {
 	var data []string
-	var stats []string
 
 	err := store.WalkDir(ctx,
-		&storage.WalkOption{SubDir: subDir},
+		&storage.WalkOption{},
 		func(path string, size int64) error {
-			// path example: /subtask/0_stat/0
-
-			// extract the parent dir
+			// extract the first dir
 			bs := hack.Slice(path)
-			lastIdx := bytes.LastIndexByte(bs, '/')
-			secondLastIdx := bytes.LastIndexByte(bs[:lastIdx], '/')
-			parentDir := path[secondLastIdx+1 : lastIdx]
+			firstIdx := bytes.IndexByte(bs, '/')
+			if firstIdx == -1 {
+				return nil
+			}
```

Comment:
how about use `strings.SplitN`

Reply:
we don't need dir info except the first 2